File: input.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
// for formatting purpose
#include <iomanip>

using namespace std;

// print non -1 values in int array
void printIntArray(int[], const int);
void printVectorPuzzle(const vector<int> &, const int);
bool checkVal(int[], const int, const int);

int main() {
    int n = 0;
    cout << "Taking input for n x n puzzle. \n\n\tFirst, please enter the size of n: ";
    cin >> n;

    cout << "\nTaking input for " << n << " x " << n << " puzzle." << endl;

    const int SIZE = n * n;
    int unusedNums[SIZE];
    cout << "\nUnused Numbers: { ";
    for(int i = 0; i < SIZE; ++i) {
        unusedNums[i] = i;
    }
    printIntArray(unusedNums, SIZE);
    cout << "}" << endl;

    int column = 1;
    int row = 1;
    int val = 0;
    vector<int> puzzle;
    cout << "\nNow, please enter" << endl;
    while (column <= n && row <= n) {
        cout << "Available options: ";
        printIntArray(unusedNums, SIZE);
        cout << endl;

        cout << "\trow: [" << row << "], column: [" << column << "] : ";
        while(cin >> val) { 
            if(checkVal(unusedNums, SIZE, val)) {
                break;
            }

            cout << "{ " << val << " } already exist or is not valid option\n" << "\tTry again: ";
        }
        
        puzzle.push_back(val);

       
        column++;
        if(row < n && column > n) {
            row++;
            column = 1;
        }
    }

    cout << endl << "result:" << endl;
    printVectorPuzzle(puzzle, n);
    cout << endl;
    

    return 0;
} 


void printIntArray(int arr[], const int SIZE) {
    for(int i = 0; i < SIZE; ++i) {
        if(arr[i] != -1) {
            cout << arr[i] << " ";
        }
    }

    return;
}

// my original code
// void printVectorPuzzle(const vector<int> &puzzle, const int N) {
//     int counter = 1;

//     for(int i = 0; i < puzzle.size(); ++i) {
//         cout << puzzle.at(i) << " ";
//         counter++;
//         if(counter > N) {
//             cout << endl;
//             counter = 1;
//         }
//     }

//     return;
// }


// getting help with formatting from outside library and LM's
void printVectorPuzzle(const vector<int> &puzzle, const int N) {
    // Find width based on max possible number (N*N - 1)
    int maxNum = N * N - 1;
    int width = to_string(maxNum).length() + 1; // +1 for spacing

    for (int i = 0; i < puzzle.size(); ++i) {
        cout << setw(width) << puzzle.at(i);
        if ((i + 1) % N == 0)
            cout << '\n';
    }

    return; 
}

bool checkVal(int arr[], const int SIZE, const int VAL) {
    for(int i = 0; i < SIZE; ++i) {
        // cout << "comparing ( " << arr[i] << ", " << VAL << " )" << endl;
        if(arr[i] == VAL) {
            arr[i] = -1;
            return true;
        }
    }

    return false;
}

File: input.py
from typing import List
from misplacedH import getMisplacedHuristic


class PuzzleInput:
    """Handles user input and construction of an n x n puzzle grid. """

    def __init__(self):
        self.n = 0
        self.puzzle: List[List[int]] = []
        self.unused_nums: List[int] = []
        self.goal_state: List[List[int]] = []

    def print_int_list(self) -> None:
        """Print available numbers (non -1)."""
        print(" ".join(str(x) for x in self.unused_nums if x != -1), end=" ")

    def print_puzzle(self) -> None:
        """Display puzzle as a formatted grid."""
        if not self.puzzle:
            print("(empty puzzle)")
            return

        max_num = self.n * self.n - 1
        width = len(str(max_num)) + 1  # For alignment

        for row in self.puzzle:
            print("".join(f"{val:>{width}}" for val in row))
        print()

    def print_goal_state(self) -> None:
        """Display goal state."""
        if not self.goal_state:
            print("(goal state not defined)")
            return

        max_num = self.n * self.n - 1
        width = len(str(max_num)) + 1  # For alignment

        for row in self.goal_state:
            print("".join(f"{val:>{width}}" for val in row))
        print()

    def check_val(self, val: int) -> bool:
        """Check if value is unused; mark as used (-1) if valid."""
        for i in range(len(self.unused_nums)):
            if self.unused_nums[i] == val:
                self.unused_nums[i] = -1
                return True
        return False

    def take_input(self) -> None:
        """Prompt user to enter puzzle size and fill it interactively."""
        print("Taking input for n x n puzzle.\n")
        self.n = int(input("\tFirst, please enter the size of n: "))

        print(f"\nTaking input for {self.n} x {self.n} puzzle.")
        size = self.n * self.n

        # define goal state according to n
        self.goal_state = [[0 for _ in range(self.n)] for _ in range(self.n)]
        count = 1
        for row in range(self.n):
            for col in range(self.n):
                self.goal_state[row][col] = count
                count += 1
        self.goal_state[self.n - 1][self.n - 1] = 0

        self.unused_nums = [i for i in range(size)]
        print("\nUnused Numbers: { ", end="")
        self.print_int_list()
        print("}")

        # initialize puzzle with zeros
        self.puzzle = [[0 for _ in range(self.n)] for _ in range(self.n)]

        print("\nNow, please enter:")

        for row in range(self.n):
            for col in range(self.n):
                print("Available options: ", end="")
                self.print_int_list()
                print()

                while True:
                    try:
                        val = int(input(f"\trow: [{row + 1}], column: [{col + 1}] : "))
                        if self.check_val(val):
                            break
                        else:
                            print(f"{{ {val} }} already exists or is not a valid option.\n\tTry again: ", end="")
                    except ValueError:
                        print("\tInvalid input, please enter an integer.")

                self.puzzle[row][col] = val
                print()
                self.print_puzzle()

        print("\nResult:")
        self.print_puzzle()

        print("\nGoal:")
        self.print_goal_state()
        
    def get_n(self) -> int:
        """Return n value of the puzzle."""
        return self.n

    def get_puzzle(self) -> List[List[int]]:
        """Return the 2D puzzle list."""
        return self.puzzle

    def get_goal(self) -> List[List[int]]:
        """Return the 2D goal_state list."""
        return self.goal_state


# in case this is imported in other python file
if __name__ == "__main__":
    builder = PuzzleInput()
    builder.take_input()
    print("misplaced huristics " + str(getMisplacedHuristic(builder.get_puzzle(), builder.get_goal(), builder.get_n())))
